src/Mooltipass/MPBLEFreeAddressProvider.cpp

```diff
@@ -0,0 +1,144 @@
+{
+    if (0 == m_parentNodeNeeded && 0 == m_childNodeNeeded)
+    {
+        qDebug() << "No free address is required";
+        return;
+    }
+
+    auto addressPackage = addressFrom;
+    auto freeAddrNum = MAX_FREE_ADDR_REQ;
+    FreeAddressInfo addressInfo;
+    addressInfo.parentNodeRequested = getNodeAskedNumber(MPNode::NodeParent, freeAddrNum);
+    addressInfo.childNodeRequested = getNodeAskedNumber(MPNode::NodeChild, freeAddrNum);
+    addressPackage.append(bleProt->toLittleEndianFromInt(addressInfo.parentNodeRequested));
+    addressPackage.append(bleProt->toLittleEndianFromInt(addressInfo.childNodeRequested));
+
+    addressInfo.startingPosition = 0;
+    jobs->append(createGetFreeAddressPackage(jobs, cbProgress, addressInfo, addressPackage));
+}
+
+void MPBLEFreeAddressProvider::cleanFreeAddresses()
+{
+    m_parentNodeNeeded = 0;
+    m_parentAddrMapping.clear();
+    m_childNodeNeeded = 0;
+    m_childAddrMapping.clear();
+}
+
+quint16 MPBLEFreeAddressProvider::getNodeAskedNumber(MPNode::NodeType nodeType, int &freeAddrNum)
+{
+    auto& m_nodeNeeded = MPNode::NodeParent == nodeType ? m_parentNodeNeeded : m_childNodeNeeded;
+    if (freeAddrNum)
+    {
+        if (m_nodeNeeded > freeAddrNum)
+        {
+            int nodeNeededNum = freeAddrNum;
+            m_nodeNeeded -= freeAddrNum;
+            freeAddrNum = 0;
+            return nodeNeededNum;
+        }
+        else
+        {
+            int nodeNeededNum = m_nodeNeeded;
+            m_nodeNeeded = 0;
+            freeAddrNum -= nodeNeededNum;
+            return nodeNeededNum;
+        }
+    }
+    return 0;
+}
+
+void MPBLEFreeAddressProvider::processReceivedAddrNumber(MPNode::NodeType nodeType, const QByteArray &receivedAddr, int &pos)
+{
+    auto& m_addrMapping = MPNode::NodeParent == nodeType ? m_parentAddrMapping : m_childAddrMapping;
+    for (auto& ba : m_addrMapping)
+    {
+        if (ba.isEmpty())
+        {
+            ba.append(receivedAddr.mid(pos, MPNode::ADDRESS_LENGTH));
+            pos += MPNode::ADDRESS_LENGTH;
+        }
+    }
+}
+
+void MPBLEFreeAddressProvider::loadRemainingFreeAddresses(AsyncJobs *jobs, const QByteArray &addressFrom, const MPDeviceProgressCb &cbProgress, bool isLastChild)
+{
+    auto freeAddrNum = MAX_FREE_ADDR_REQ - 1;
+    auto addressPackage = addressFrom;
+    FreeAddressInfo addressInfo;
+    addressInfo.parentNodeRequested = getNodeAskedNumber(MPNode::NodeParent, freeAddrNum);
+    addressInfo.childNodeRequested = getNodeAskedNumber(MPNode::NodeChild, freeAddrNum);
+    if (isLastChild)
+    {
+        ++addressInfo.childNodeRequested;
+    }
+    else
+    {
+        ++addressInfo.parentNodeRequested;
+    }
+    addressPackage.append(bleProt->toLittleEndianFromInt(addressInfo.parentNodeRequested));
+    addressPackage.append(bleProt->toLittleEndianFromInt(addressInfo.childNodeRequested));
+    addressInfo.startingPosition = MPNode::ADDRESS_LENGTH;
```

Comment:
so with this variable you skip the first address already provided.
You really need to comment that code.